### PR TITLE
CP-18185 new plugin discovery mechanism

### PR DESCRIPTION
**Pull request by lindig**
I'm inviting comments for the implementation of a plugin discovery mechanism that is part of PVS. 
- For this to build I'd need to change the corresponding spec file such that `ocaml-inofify` is included. However, do we use branches for the spec file repositories?
- I still need to test this. (inotify does not work inside Docker.)
- What should happen to the existing registration mechanism? One way to solve this is to keep it and to have the new mechanism look for plugins in a different directory than `/dev/shm/metrics` (`Rrdd_server.Plugin.base_path`).

### CODE REVIEW

File: rrdd/xcp_rrdd.ml [hunk 1]
```diff
@@ -615,6 +695,13 @@ let _ =
                 with e ->
                         error "monitoring loop thread has failed" in
 
+	let () = 
+		try Discover.start () |> ignore
+		with e -> error 
+			"RRD plugin discovery thread failed: %s"
```
Comment (robhoes) [inline rrdd/xcp_rrdd.ml]: Let's restart the thread on failure (after a short delay).

**Comment by robhoes** [comment]
Looks pretty good to me so far!

**Comment by lindig** [comment]
The code works now when built against `ocaml-inotify 2.3` and tested on a DT box. A remaining problem is that RRD files created by `rrd-client-lib` can't be processed and lead to exceptions - most likely because of format incompatibilities. I've submitted a pull request for `ocaml-inotify 2.3`. To see the full logs, one needs to edit `/ect/xcp-rrdd.conf`.

**Comment by lindig** [comment]
The code now works and successfully registers and un-registers plugins on a DT machine. However, the mechanism is still a bit fragile for a different reason: because data is immediately read when a plugin is registered, a plugin cannot create the file and write content only later. It has to provide valid content as it creates the file. This is currently not the case for files created by rrd-client-lib. This library creates a file and updates it later, leading to read errors in xcp-rrdd.

**Comment by lindig** [comment]
The remaining problem of plugin registration being fragile is addressed by changes on the client side at https://github.com/lindig/rrd-client-lib-1/tree/rrdd-protocol. The client now writes a file of constant size and makes sure the content is valid.

**Comment by jonludlam** [comment]
There's a lot of code duplicated between the standalone util and rrdd - if we're intending to keep the stand-alone tool around we should dedup, and if not, we should rm it.

**Comment by lindig** [comment]
That is a fair point. I was using it for experiments with the inotify mechanism. It is more a tool than an integrated part.

File: rrdd/xcp_rrdd.ml [hunk 2]
```diff
@@ -538,6 +538,121 @@ let monitor_loop () =
...
+			( Inotify.add_watch fd dir selectors |> ignore
+			; loop []
+			)
+
+	(* [scan] scans a directory for plugins and registers them *)
+	let scan dir =
+		debug "RRD plugin - scanning %s" dir;
+		Sys.readdir dir
+		|> Array.to_list
+		|> List.iter register
```
Comment (jonludlam) [inline rrdd/xcp_rrdd.ml:644]: Do you need a filter by is_valid here?

Comment (lindig) [reply rrdd/xcp_rrdd.ml:644]: Yes - I should add it. The expectation is that these tmp files are very short lived and don't exist at startup time. But they could exist when we recover from a crash. In any case, we should always ignore tmp files.

**Comment by jonludlam** [comment]
Cool, looks good now.

**Pull request merged**